central/resourcecollection/service/service_impl.go

```diff
@@ -83,3 +94,64 @@ func (s *serviceImpl) getCollection(ctx context.Context, id string) (*v1.GetColl
 		Deployments: nil,
 	}, nil
 }
+
+// CreateCollection creates a new collection from the given request
+func (s *serviceImpl) CreateCollection(ctx context.Context, request *v1.CreateCollectionRequest) (*v1.CreateCollectionResponse, error) {
+	if !features.ObjectCollections.Enabled() {
+		return nil, errors.New("Resource collections is not enabled")
+	}
+
+	collection, err := collectionRequestToCollection(ctx, request, true, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	err = s.datastore.AddCollection(ctx, collection)
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.CreateCollectionResponse{Collection: collection}, nil
+}
+
+func collectionRequestToCollection(ctx context.Context, request collectionRequest, isCreate bool, getID func() string) (*storage.ResourceCollection, error) {
+	if request.GetName() == "" {
+		return nil, errors.Wrap(errox.InvalidArgs, "Collection name should not be empty")
+	}
+
+	slimUser := utils.UserFromContext(ctx)
+	if slimUser == nil {
+		return nil, errors.New("Could not determine user identity from provided context")
+	}
+
+	if len(request.GetResourceSelectors()) == 0 {
+		return nil, errors.Wrap(errox.InvalidArgs, "No resource selectors were provided")
+	}
+
+	timeNow := protoconv.ConvertTimeToTimestamp(time.Now())
+
+	collection := &storage.ResourceCollection{
+		Name:              request.GetName(),
+		Description:       request.GetDescription(),
+		LastUpdated:       timeNow,
+		UpdatedBy:         slimUser,
+		ResourceSelectors: request.GetResourceSelectors(),
+	}
+
+	if isCreate {
+		collection.CreatedBy = slimUser
+		collection.CreatedAt = timeNow
+	} else if getID != nil {
+		collection.Id = getID()
+	}
```

Comment:
implying `getID` wrt `isCreate` is confusing. Strongly recommend dropping this check and explicitly basing the condition on create and update workflow.

Reply:
Ah, I merged this PR before noticing the comment. I will add it in my Delete collections service PR. Should I change to something like below?
```
if isCreate {...}
else {
    if getID == nil {
        return nil, err
    }
    collection.Id = getID()
}
```